### PR TITLE
Microbial

### PR DESCRIPTION
**Pull request by robinandeer**
Support submission of microbial samples

- [x] write script to migrate database, add new tables: microbial_sample, order

### CODE REVIEW

**Comment by ingkebil** [comment]
hehe, too bad you merged master into this branch as now the broken master commits are here too ;) Let's see if we can fix that

**Comment by ingkebil** [comment]
Removed the broken-master commits

**Comment by ingkebil** [comment]
... aaaaand, I broke it

**Comment by ingkebil** [comment]
... OK I think I found the PR ;)

File: cg/server/api.py [hunk 1]
```diff
@@ -150,6 +150,60 @@ def sample(sample_id):
...
 
 
+@BLUEPRINT.route('/microbial_orders')
+def microbial_orders():
+    """Fetch microbial orders."""
+    customer_obj = None if g.current_user.is_admin else g.current_user.customer
+    orders_q = db.microbial_orders(
+        query=request.args.get('query'),
+        customer=customer_obj,
+        action=request.args.get('action'),
```
Comment (ingkebil) [inline cg/server/api.py]: what does the `action=request.args.get('action')` do?

File: cg/server/api.py [hunk 2]
```diff
@@ -150,6 +150,60 @@ def sample(sample_id):
...
+    limit = int(request.args.get('limit', 50))
+    data = [sample_obj.to_dict(order=True) for sample_obj in samples_q.limit(limit)]
+    return jsonify(samples=data, total=samples_q.count())
+
+
+@BLUEPRINT.route('/microbial_samples/<sample_id>')
+def microbial_sample(sample_id):
+    """Fetch a single sample."""
+    sample_obj = db.microbial_sample(sample_id)
+    print(sample_id)
```
Comment (ingkebil) [inline cg/server/api.py]: stray print

File: cg/store/api/add.py [hunk 3]
```diff
@@ -137,15 +136,50 @@ def add_delivery(self, destination: str, sample: models.Sample=None, pool: model
...
-                    pools: List[models.Pool]=None, comment: str=None, discount: int=0, 
-                    record_type: str=None):
+    def add_invoice(self, customer: models.Customer, samples: List[models.Sample] = None,
+                    pools: List[models.Pool] = None, comment: str = None, discount: int = 0,
+                    record_type: str = None):
         """Build a new Invoice record."""
         new_id = self.new_invoice_id()
-        new_invoice = self.Invoice(comment=comment, discount=discount, id=new_id, record_type=record_type)
+        new_invoice = self.Invoice(comment=comment, discount=discount, id=new_id,
+                                   record_type=record_type)
```
Comment (ingkebil) [inline cg/store/api/add.py]: reviewed until here

File: cg/server/admin.py [hunk 4]
```diff
@@ -114,4 +114,23 @@ class AnalysisView(BaseView):
...
+
+
+class MicrobialSampleView(BaseView):
+    column_searchable_list = ['internal_id', 'name', 'microbial_order.ticket_number']
+    column_editable_list = ['reads', 'comment', 'reference_genome']
+    column_filters = ['microbial_order', 'microbial_order.customer']
+    column_default_sort = ('created_at', True)
+
+
+class InvoiceView(BaseView):
```
[the remark this answers is not in the document]
Comment (patrikgrenfeldt) [reply cg/server/admin.py]: Removed

File: cg/store/api/add.py [hunk 5]
```diff
@@ -137,15 +136,50 @@ def add_delivery(self, destination: str, sample: models.Sample=None, pool: model
...
+                             application_version: models.ApplicationVersion,
+                             priority: str = None,
+                             comment: str = None, **kwargs) -> models.MicrobialSample:
+        """Build a new MicrobialSample record.
+        
+        To commit you also need to assign the sample to an Order.
+        """
+        internal_id = internal_id or utils.get_unique_id(self.sample)
+        priority_human = priority or 'standard'
+        db_priority = PRIORITY_MAP[priority_human]
```
Comment (patrikgrenfeldt) [inline cg/store/api/add.py]: db_priority = PRIORITY_MAP[priority_human] 
make the priorities consistent regarding order

File: cg/store/api/find.py [hunk 6]
```diff
@@ -59,6 +60,20 @@ def samples(self, *, customer: models.Customer=None, query: str=None) -> List[mo
...
         return records.order_by(models.Sample.created_at.desc())
 
+    def microbial_samples(self, *, customer: models.Customer = None, query: str = None) -> List[
+        models.MicrobialSample]:
+        records = self.MicrobialSample.query
+        records = records.filter_by(customer=customer) if customer else records
+        records = records.filter(or_(
+            models.MicrobialSample.name.like(f"%{query}%"),
+            models.MicrobialSample.internal_id.like(f"%{query}%"),
+        )) if query else records
```
Comment (patrikgrenfeldt) [inline cg/store/api/find.py]: rename query as it is an sqlalchemy keyword. Good luck!

File: cg/store/api/find.py [hunk 7]
```diff
@@ -91,17 +106,18 @@ def panel(self, abbrev):
...
-                join(models.Analysis.family).\
-                filter(models.Analysis.started_at < before).\
-                group_by(models.Family.id).\
-                with_entities(models.Analysis.family_id, func.max(models.Analysis.started_at).label('started_at')).subquery()
+            subq = self.Analysis.query. \
+                join(models.Analysis.family). \
+                filter(models.Analysis.started_at < before). \
+                group_by(models.Family.id). \
+                with_entities(models.Analysis.family_id,
+                              func.max(models.Analysis.started_at).label('started_at')).subquery()
```
Comment (patrikgrenfeldt) [inline cg/store/api/find.py]: remove empty line

File: cg/store/models.py [hunk 8]
```diff
@@ -162,6 +162,86 @@ def panels(self, panel_list: List[str]):
...
+    microbial_order_id = Column(ForeignKey('microbial_order.id'), nullable=False)
+    created_at = Column(types.DateTime, default=dt.datetime.now)
+    updated_at = Column(types.DateTime, onupdate=dt.datetime.now)
+    received_at = Column(types.DateTime)
+    prepared_at = Column(types.DateTime)
+    sequence_start = Column(types.DateTime)
+    sequenced_at = Column(types.DateTime)
+    delivered_at = Column(types.DateTime)
+    strain = Column(types.String(32))
+    strain_other = Column(types.String(32))
```
Comment (patrikgrenfeldt) [inline cg/store/models.py]: put both strain and strain_other to length 255

File: cg/store/models.py [hunk 9]
```diff
@@ -162,6 +162,86 @@ def panels(self, panel_list: List[str]):
...
+    reference_genome = Column(types.String(32))
```
Comment (patrikgrenfeldt) [inline cg/store/models.py]: as well as this one

File: cg/store/models.py [hunk 10]
```diff
@@ -162,6 +162,86 @@ def panels(self, panel_list: List[str]):
...
+        elif self.received_at:
+            return f"Received {self.received_at.date()}"
+        else:
+            return f"Ordered {self.ordered_at.date()}"
+
+    def to_dict(self, order=False) -> dict:
+        """Override dictify method."""
+        data = super(MicrobialSample, self).to_dict()
+        data['application_version'] = self.application_version.to_dict()
+        data['application'] = self.application_version.application.to_dict()
```
Comment (patrikgrenfeldt) [inline cg/store/models.py]: do we use both of these?
is application version not always available through application?

Comment (patrikgrenfeldt) [reply cg/store/models.py]: does this list the application and application version the sample was ordered with?

File: cg/store/api/find.py (hunk 6 above)
Comment (patrikgrenfeldt) [reply cg/store/api/find.py]: Done

**Pull request merged**